dep/chromium/base/numerics/safe_conversions_impl.h

```diff
@@ -0,0 +1,850 @@
+                           std::is_signed<NumericType>::value;
+};
+
+// Helper templates for integer manipulations.
+
+template <typename Integer>
+struct PositionOfSignBit {
+  static const size_t value = IntegerBitsPlusSign<Integer>::value - 1;
+};
+
+// Determines if a numeric value is negative without throwing compiler
+// warnings on: unsigned(value) < 0.
+template <typename T,
+          typename std::enable_if<std::is_signed<T>::value>::type* = nullptr>
+constexpr bool IsValueNegative(T value) {
+  static_assert(std::is_arithmetic<T>::value, "Argument must be numeric.");
+  return value < 0;
+}
+
+template <typename T,
+          typename std::enable_if<!std::is_signed<T>::value>::type* = nullptr>
+constexpr bool IsValueNegative(T) {
+  static_assert(std::is_arithmetic<T>::value, "Argument must be numeric.");
+  return false;
+}
+
+// This performs a fast negation, returning a signed value. It works on unsigned
+// arguments, but probably doesn't do what you want for any unsigned value
+// larger than max / 2 + 1 (i.e. signed min cast to unsigned).
+template <typename T>
+constexpr typename std::make_signed<T>::type ConditionalNegate(
+    T x,
+    bool is_negative) {
+  static_assert(std::is_integral<T>::value, "Type must be integral");
+  using SignedT = typename std::make_signed<T>::type;
+  using UnsignedT = typename std::make_unsigned<T>::type;
+  return static_cast<SignedT>(
+      (static_cast<UnsignedT>(x) ^ -SignedT(is_negative)) + is_negative);
+}
+
+// This performs a safe, absolute value via unsigned overflow.
+template <typename T>
+constexpr typename std::make_unsigned<T>::type SafeUnsignedAbs(T value) {
+  static_assert(std::is_integral<T>::value, "Type must be integral");
+  using UnsignedT = typename std::make_unsigned<T>::type;
+  return IsValueNegative(value) ? 0 - static_cast<UnsignedT>(value)
+                                : static_cast<UnsignedT>(value);
+}
+
+// This allows us to switch paths on known compile-time constants.
+#if defined(__clang__) || defined(__GNUC__)
+constexpr bool CanDetectCompileTimeConstant() {
+  return true;
+}
+template <typename T>
+constexpr bool IsCompileTimeConstant(const T v) {
+  return __builtin_constant_p(v);
+}
+#else
+constexpr bool CanDetectCompileTimeConstant() {
+  return false;
+}
+template <typename T>
+constexpr bool IsCompileTimeConstant(const T) {
+  return false;
+}
+#endif
+template <typename T>
+constexpr bool MustTreatAsConstexpr(const T v) {
+  // Either we can't detect a compile-time constant, and must always use the
+  // constexpr path, or we know we have a compile-time constant.
+  return !CanDetectCompileTimeConstant() || IsCompileTimeConstant(v);
+}
+
+// Forces a crash, like a CHECK(false). Used for numeric boundary errors.
+// Also used in a constexpr template to trigger a compilation failure on
+// an error condition.
+struct CheckOnFailure {
+  template <typename T>
+  static T HandleFailure() {
+#if defined(_MSC_VER)
+    __debugbreak();
```

Comment:
This is the only line which you need to keep in mind when integrating this code IMO.
AFAICS `SafeInt` only throws an exception in case of a failure, whereas this library causes a hard crash.
You could modify the code to throw an exception instead - for instance a primitive subclass of `std::range_error`. 🙂

Reply:
I honestly didn't read every line of this before integrating it here. I presumed that one of the 4 classes of math here would throw on failure instead of crashing hard or that there was a template argument to override the behavior. Is that not true? If we need to modify this, we will. I was hoping to get this in, start using it, then modify it only if necessary when we discover problems.

Reply:
> I presumed that one of the 4 classes of math here would throw on failure instead of crashing hard

I hadn't looked at it in detail either, but I think you're right. Based on what they say in the README, I think the "checked" APIs crash on error, and the "strict" ones throw.

Reply:
This will not really be an issue in the beginning @miniksa @j4james. 🙂
Only 2 parts of the numerics API will call `HandleFailure()` and "crash" the process:
* `CheckedNumeric`'s `ValueOrDie` - I suppose the naming of the function is very obvious 😄
* `checked_cast<Dst, CheckHandler = CheckOnFailure, Src>()` - we can just use `gsl::narrow_cast` instead, or simply create an alias for it in the `til` namespace with a replacement for the `CheckHandler`

Either way we should really only be careful about `checked_cast` at most IMO.
Also "crashing" during an erroneous checked cast is actually not even that bad of an idea in my opinion. 😶
I just wanted to mention this circumstance because I know that `gsl::narrow_cast` is being used, which throws an exception instead.